src/models/room.ts

```diff
@@ -2249,6 +2288,184 @@ export class Room extends EventEmitter {
             return "Empty room";
         }
     }
+
+    /**
+     * When we receive a new visibility change event:
+     *
+     * - store this visibility change alongside the timeline, in case we
+     *   later need to apply it to an event that we haven't received yet;
+     * - if we have already received the event whose visibility has changed,
+     *   patch it to reflect the visibility change and inform listeners.
+     */
+    private applyNewVisibilityChangeEvent(event: MatrixEvent): void {
+        // Perform sanity checks on the event.
+        if (!event.isEventVisibilityChange()) {
+            // Internal error.
+            throw new Error("Expected a visibility change relation");
+        }
+
+        const relation = event.getRelation();
+        const originalEventId = relation.event_id;
+        let visible: boolean;
+        switch (event.getEventVisibilityChange()) {
+            case "hidden":
+                visible = false;
+                break;
+            case "visible":
+                visible = true;
+                break;
+            default:
+                // Event is ill-formed.
+                return;
+        }
+        const reason = visible ? event.getContent()["reason"] : null;
+        if (reason != null && reason !== undefined && typeof reason != "string") {
+            // Event is ill-formed.
+            return;
+        }
+
+        // Ignore visibility change events that are not emitted by moderators.
+        const powerLevelsEvents = this.currentState.getStateEvents(EventType.RoomPowerLevels, "");
+        const powerLevels = powerLevelsEvents && powerLevelsEvents.getContent();
+        if (!powerLevels || powerLevels["org.matrix.msc3531.visibility"] < event.sender.powerLevel) {
+            return;
+        }
+
+        // Record this change in visibility.
+        // If the event is not in our timeline and we only receive it later,
+        // we may need to apply the visibility change at a later date.
+
+        const visibilityChangesOnOriginalEvent = this.visibilityChanges.get(originalEventId);
+        if (visibilityChangesOnOriginalEvent) {
+            // It would be tempting to simply erase the latest visibility change
+            // but we need to record all of the changes in case the latest change
+            // is ever redacted.
+            // In practice, it should be unusual for an event to have more than
+            // 2 visibility changes, so linear scans through `visibilityChanges`
+            // should be fast. It might be possible to DoS this feature, though.
```

Comment:
We can always apply a limit to the loop, and we probably should if it's highlighted as an attack here.

Reply:
How should we react when we reach the limit?

Reply:
I'd just truncate. If someone is flipping visibility more than 50 times, they kinda deserve whatever state we ended up on.